lib/msal-browser/src/app/PublicClientApplication.ts

```diff
@@ -114,8 +114,19 @@ export class PublicClientApplication {
         const cachedHash = this.browserStorage.getItem(TemporaryCacheKeys.URL_HASH);
         try {
             // If hash exists, handle in window. Otherwise, cancel any current requests and continue.
-            const interactionHandler = new RedirectHandler(this.authModule, this.browserStorage, this.config.auth.navigateToLoginRequestUrl);
-            const responseHash = UrlString.hashContainsKnownProperties(hash) ? hash : cachedHash;
+            const interactionHandler = new RedirectHandler(this.authModule, this.browserStorage);
+            let responseHash: string;
+            if (this.config.auth.navigateToLoginRequestUrl && UrlString.hashContainsKnownProperties(hash)) {
+                this.browserStorage.setItem(TemporaryCacheKeys.URL_HASH, hash);
+                interactionHandler.navigateToRequestUrl();
+                return;
+            } else if (!this.config.auth.navigateToLoginRequestUrl) {
+                responseHash = UrlString.hashContainsKnownProperties(hash) ? hash : cachedHash;
+                window.location.hash = "";
+            } else {
+                responseHash = cachedHash;
+            }
```

Comment:
This block of code is confusing to understand, given the `if/else if/else` and mutation. From our discussion, each of these is for a different scenario, but it is hard to deduce that just from reading the code. My suggestions:

1. Regardless of changes to syntax, add comments making clear what is happening.
2. Avoid mutation if possible (not all mutation is bad, but I personally treat it as the exception, not the rule).
3. Similar to redirect, I think there should be a helper function to modifying `window.location.hash`.
4. Refactor code as follows:

```js
const validHash = UrlString.hashContainsKnownProperties(hash);
// Scenario 1
if (this.config.auth.navigateToLoginRequestUrl && validHash) {
  this.browserStorage.setItem(TemporaryCacheKeys.URL_HASH, hash);
  const loginRequestUrl = this.browserStorage.getItem(TemporaryCacheKeys.ORIGIN_URI);
  interactionHandler.navigateToRequestUrl(loginRequestUrl);
  return;
}

// Scenario 2
if (!this.config.auth.navigateToLoginRequestUrl) {
   window.location.hash = ""; // factor out into helper
}

// Scenario 2 / 3
const responseHash = validHash ? hash : cachedHash;
```

Reply:
I've removed the if/else mutation and added some helpers instead. Please take a look.